File: 20.py
from typing import List, Optional
import sys

class Solution:
    def isValid(self, s: str) -> bool:
        match = {")": "(", "]": "[", "}": "{"}
        stack = []
        for c in s:
            if len(stack) == 0:
                stack.append(c)
            elif c in match and match[c] == stack[-1]:
                del stack[-1]
            else:
                stack.append(c)
        return len(stack) == 0 

if __name__ == '__main__':
    s = r"()[]{}"
    sol = Solution()
    print(sol.isValid(s))

File: leetcode_cpp/20.cpp
/*
有效的括号 栈
给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。
有效字符串需满足：
左括号必须用相同类型的右括号闭合。
左括号必须以正确的顺序闭合。
每个右括号都有一个对应的相同类型的左括号。
*/
using namespace std;
#include<vector>
#include<iostream>
#include<algorithm>
#include<numeric>
#include<unordered_map>
#include<math.h>
#include<queue>
#include<stack>
#include<functional>

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        unordered_map<char, char> match;
        match[')'] = '(';
        match[']'] = '[';
        match['}'] = '{';
        for(auto c : s) {
            if(st.empty()) {
                st.push(c);
            }
            else if(match.find(c) != match.end() && match[c] == st.top()) {
                st.pop();
            }
            else {
                st.push(c);
            }
        }
        return st.empty();
    }
};

int main()
{
    string s = "()[]{}";
	Solution sol;
    cout << sol.isValid(s) << endl;
	system("pause");
}


